// icloud-shared-album/build/redirectedBaseUrl.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRedirectedBaseUrl = void 0;
// 2024 : Apple began issuing a 330 redirect, so we need to deal with that.
// With thanks to:
// https://github.com/bertrandom/icloud-shared-album-to-flickr/commit/badd91d53b96832213c9e156a31ad604aa740014#diff-e07d531ac040ce3f40e0ce632ac2a059d7cd60f20e61f78268ac3be015b3b28f
// Adapted by @jmaton : use axios instead of reqeust-promise-native; massage syntax to match this project (ICloud-Shared-Album)
var axios_1 = require("axios");
var getRedirectedBaseUrl = function (baseUrl, token) { return __awaiter(void 0, void 0, void 0, function () {
    var url, headers, dataString, response, body, newBaseUrl;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                url = baseUrl + 'webstream';
                headers = {
                    'Origin': 'https://www.icloud.com',
                    'Accept-Language': 'en-US,en;q=0.8',
                    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36',
                    'Content-Type': 'text/plain',
                    'Accept': '*/*',
                    'Referer': 'https://www.icloud.com/sharedalbum/',
                    'Connection': 'keep-alive'
                };
                dataString = '{"streamCtag":null}';
                return [4 /*yield*/, (0, axios_1.default)({
                        url: url,
                        method: 'POST',
                        headers: headers,
                        data: dataString,
                        validateStatus: function (status) { return status < 400; } // do not fail on 330
                    })];
            case 1:
                response = _a.sent();
                if (response.status == 330) {
                    body = response.data;
                    newBaseUrl = 'https://';
                    newBaseUrl += body["X-Apple-MMe-Host"];
                    newBaseUrl += '/';
                    newBaseUrl += token;
                    newBaseUrl += '/sharedstreams/';
                    return [2 /*return*/, newBaseUrl];
                }
                return [2 /*return*/, baseUrl];
        }
    });
}); };
exports.getRedirectedBaseUrl = getRedirectedBaseUrl;
